emu_mps/results.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+from pulser.backend import Results
+import json
+from typing import cast, Any
+import numpy as np
+import torch
+from uuid import UUID
+from copy import deepcopy
+
+
+class AbstractReprEncoder(json.JSONEncoder):
+    """The custom encoder for abstract representation of Pulser objects."""
+
+    def default(self, o: Any) -> dict[str, Any] | list | int | float | str:
+        """Handles JSON encoding of objects not supported by default."""
+        if hasattr(o, "_to_abstract_repr"):
+            return cast(dict, o._to_abstract_repr())
+        elif isinstance(o, np.ndarray):
+            return cast(list, o.tolist())
+        elif isinstance(o, torch.Tensor):
+            if len(o.shape) == 0:
+                return o.item()
+            else:
+                return o.tolist()
+        elif isinstance(o, UUID):
+            return str(o)
+        elif isinstance(o, np.integer):
+            return int(o)
+        elif isinstance(o, set):
+            return list(o)
+        elif isinstance(o, complex):
+            return dict(real=o.real, imag=o.imag)
+        else:  # pragma: no cover
+            return cast(dict, json.JSONEncoder.default(self, o))
+
+
+class MPSResults(Results):
+    def _to_abstract_repr(self) -> dict:
+        d = deepcopy(self.__dict__)
+        for key, value in d["_tagmap"].items():
+            d["_tagmap"][key] = str(value)
+        tmp = {}
+        for key, value in d["_results"].items():
+            tmp[str(key)] = value
+        d["_results"] = tmp
```

Comment:
using dict comprehension is more readable and doesn't require extra `tmp`:
`d["_results"] = {str(key): value for key, value in d["_results"].items()}`

applies to all loops in this function:
`d["_tagmap"] = {k: str(v) for k, v in d["_tagmap"].items()}`

Reply:
done